clog/uwsgi_plugin.py

```diff
@@ -0,0 +1,58 @@
+import uwsgi
+import uwsgidecorators
+import logging
+import clog
+import clog.global_state
+import clog.handlers
+
+try:
+    import cPickle as pickle
+except ImportError:
+    import pickle
+
+
+def _mule_msg(*args, **kwargs):
+    mule = kwargs.pop('mule', 1)
+    data = pickle.dumps({
+        'service': 'uwsgi_mulefunc',
+        'func': 'log_line',
+        'args': args,
+        'kwargs': kwargs,
+    })
+    uwsgi.mule_msg(data, mule)
+
+
+class UwsgiHandler(logging.Handler):
+
+    def __init__(self, stream, mule=1):
+        logging.Handler.__init__(self)
+        self.stream = stream
+        self.mule = mule
+
+    def emit(self, record):
+        try:
+            msg = self.format(record)
+            _mule_msg(self.stream, msg, mule=self.mule)
+        except:
+            self.handleError(record)
+
+
+# TODO: Patch upstream uwsgi to return success on the pipe write so
+# we can optionally default to sync log_line
+def uwsgi_log_line(stream, line, mule=1):
+    _mule_msg(stream, line, mule=mule)
+
+
+def uwsgi_patch_global_state():
+    map(
+        lambda x: setattr(x, 'log_line', uwsgi_log_line),
+        (clog, clog.global_state)
+    )
+
+
+# Couple setup tasks at import:
+# 1. Register 'log_line' in all mules for dispatch handling on the mule side
+# 2. Insert 'UwsgiHandler' into the clog.handlers module for seamless usage
+
+uwsgidecorators.mule_functions['log_line'] = clog.global_state.log_line
+setattr(clog.handlers, 'UwsgiHandler', UwsgiHandler)
```

Comment:
That's fair, I _do_ think it could be factored in a different way.

Here's a sketch, idk if it works but it's an idea

```
clog/
     __init__.py
     global_state.py
     _impl_uwsgi.py
     _impl_default.py
```

`__init__.py` remains unchanged

`global_state.py` remains largely unchanged except the implementation of `log_line` moves to `_default_impl.py` and the bottom of the file looks like:

```python
try:
    from ._impl_uwsgi import log_line
    from ._impl_uwsgi import ...
except ImportError:
    from ._impl_default import log_line
    from ._impl_default import ...
```

The two impl files implement their specific code

idrc either way, feel free to drop this suggestion :)